Core/COEditingContext.m

```diff
@@ -310,6 +325,105 @@ - (void)undeletePersistentRoot: (COPersistentRoot *)aPersistentRoot
     {
         [_persistentRootsPendingUndeletion addObject: aPersistentRoot];
     }
+
+	[self updateCrossPersistentRootReferencesToPersistentRoot: aPersistentRoot
+	                                                   branch: nil
+	                                                isDeleted: NO];
+	
+	/* Once all the dead references are fixed, remove them from the cache */
+	
+	NSSet * allBranches = [aPersistentRoot.branches setByAddingObjectsFromSet: aPersistentRoot.deletedBranches];
+	
+	for (COBranch *branch in allBranches)
+	{
+		[_deadRelationshipCache removePath: [COPath pathWithPersistentRoot: aPersistentRoot.UUID
+		                                                            branch: branch.UUID]];
+	}
+	[_deadRelationshipCache removePath: [COPath pathWithPersistentRoot: aPersistentRoot.UUID]];
+}
+
+/**
+ * When isDeleted is YES, turns live references to the given persistent root 
+ * into dead ones in referring persistent roots.
+ *
+ * When isDeleted is NO, turns dead references to the given persistent root
+ * into live ones in referring persistent roots.
+ *
+ * When isDeleted is YES and branch is nil, this means the persistent root 
+ * itself is deleted. In this case, we consider the branches to be all 
+ * transively deleted for this method logic, although there aren't marked as 
+ * deleted.
+ *
+ * The implementation must take in account deletion/undeletion can be:
+ *
+ * <list>
+ * <item>explicit with COPersistentRoot and COBranch API</item>
+ * <item>implicit deletion/undeletion when reloading persistent roots or branches 
+ * (e.g. isTargetDeletion comment).</item>
+ * </list>
+ *
+ * IMPORTANT: This method must never result in new changes to be committed, 
+ * since it replaces object references in memory, but doesn't change any item
+ * graphs (the references inside each COItem remain identical). Requiring no 
+ * commits means the references can be fixed in both tracking and current branches.
+ *
+ * See also -[COPersistentRoot setCurrentBranchObjectGraphToRevisionUUID:persistentRootUUID:].
+ */
+- (void)updateCrossPersistentRootReferencesToPersistentRoot: (COPersistentRoot *)aPersistentRoot
+                                                     branch: (COBranch *)aBranch
+                                                  isDeleted: (BOOL)isDeletion
+{
+	NSParameterAssert(aPersistentRoot != nil);
+	// NOTE: -delete/undeleteBranch: enforce !aBranch.isCurrentBranch already.
+	NSParameterAssert(aBranch == nil
+		|| (!aBranch.isCurrentBranch && aBranch.persistentRoot == aPersistentRoot));
+
+	for (COPersistentRoot *persistentRoot in [_loadedPersistentRoots objectEnumerator])
+	{
+		if (persistentRoot == aPersistentRoot)
+			continue;
+		
+		// TODO: Use -objectGraphWithoutUnfaulting or -allObjectGraphContexts to
+		// prevent loading every object graph contexts we check
```

Comment:
Yeah, this TODO would be good to have.

From what I can see, this method is a "traverse loaded object graphs and fix up pointers" kind, it really shouldn't do any unfaulting, for both performance and correctness.

Also, it seems to be fixing all loaded persistent roots, couldn't it use the dead reference cache/relationship cache to only fix up ones that need it?